### PR TITLE
Readdir byname

### PR DESCRIPTION
**Pull request by balajisrao**

### CODE REVIEW

File: vfs_proxyfs.c [hunk 1]
```diff
@@ -436,6 +435,16 @@ static DIR *vfs_proxyfs_fdopendir(struct vfs_handle_struct *handle,
 	return (DIR *) *(file_handle_t **)VFS_FETCH_FSP_EXTENSION(handle, fsp);
 }
 
+/*
+ * The readdir() function returns a pointer to a dirent structure representing the next directory entry
+ * in the directory stream pointed to by dirp.
+ *
+ * Note on next direntry pointer:
+ * In proxyfs we use last read name as the marker instead of the corresponding dirent.d_off. Because, the internal
+ * directory btree key is name (offset is just hte location of the entry in the directory at the time of making the call).
```
Comment (bschatz-swift) [inline vfs_proxyfs.c]: "hte" should be "the"

File: vfs_proxyfs.c [hunk 2]
```diff
@@ -436,6 +435,16 @@ static DIR *vfs_proxyfs_fdopendir(struct vfs_handle_struct *handle,
...
+ * Therefore, if files are deleted then offet corresponding to a name could change, leading to inconsistency if we use it
+ * as a marker.
+ */
```
Comment (bschatz-swift) [inline vfs_proxyfs.c]: Nice comment!!

File: vfs_proxyfs.c [hunk 3]
```diff
@@ -476,19 +497,58 @@ static struct dirent *vfs_proxyfs_readdir(struct vfs_handle_struct *handle,
...
+		dir->use_name = dir->telldir_info[1].use_name;
+		dir->prev_readdir_name = dir->telldir_info[1].name;
+	} else {
+		// We are seeking to a location for which we don't have corresponding name, we fall back to
+		// using offset as the marker, this could lead to inconsistency if files were deleted in the directory
+		// after getting the offset from telldir().
+		// This should not happen in case of samba requests!
+		DEBUG(3, ("vfs_proxyfs_seekdir: %s offset : %ld seekign to a location we are not trakcing by name, could lead to readdir inconsistency\n", handle->conn->connectpath, offset));
+		dir->offset = offset;
+		dir->use_name = false;
```
Comment (bschatz-swift) [inline vfs_proxyfs.c]: Do you want to ASSERT() here since it is an inconsistency?

File: vfs_proxyfs.h [hunk 4]
```diff
@@ -12,12 +12,21 @@ typedef struct {
...
+
 typedef struct file_handle_s {
-	uint64_t      inum;
-	off_t         offset;
-	uint64_t      flags;
-	uint64_t      mode;
-	struct dirent dir_ent;
+	uint64_t       inum;
+	off_t          offset;
+	bool		   use_name;  // Use last returned name as the marker for readdir otherwise offset will be used.
```
Comment (bschatz-swift) [inline vfs_proxyfs.h]: NIT: Line up "use_name" with the line above it...

File: vfs_proxyfs.h [hunk 5]
```diff
@@ -12,12 +12,21 @@ typedef struct {
...
+	uint64_t       flags;
+	uint64_t       mode;
+	struct dirent  dir_ent;
+	char           *prev_readdir_name;
+	telldir_info_t telldir_info[2];
```
Comment (bschatz-swift) [inline vfs_proxyfs.h]: Which is oldest marker - element 0 or element 1?

File: vfs_proxyfs.c (hunk 1 above)
Comment (balajisrao) [reply vfs_proxyfs.c]: fixed

**Pull request merged**